tests/e2e/test_spyre_cb.py

```diff
@@ -83,177 +91,412 @@ def test_cb_handling(
+
+    assert len(engine_core.scheduler.waiting) == 0
+    assert len(engine_core.scheduler.running) == 2
+    request_outputs = engine.step()
+    assert len(request_outputs) == 2  # requests 2 and 3 decoding now
+    assert request_outputs[0].request_id == "3"  # req 3 is decoding
+    assert request_outputs[1].request_id == "2"  # req 2 is decoding
+
+    assert len(engine_core.scheduler.waiting) == 0
+    assert len(engine_core.scheduler.running) == 2
+    request_outputs = engine.step()
+    assert len(request_outputs) == 2  # requests 2 and 3 decoding now
+    assert request_outputs[0].request_id == "3"  # req 3 is decoding
+    assert request_outputs[1].request_id == "2"  # req 2 is decoding
+    assert request_outputs[1].finished  # request 2 is done
+    assert request_outputs[1].outputs[0].text == " 6 5 4 3 2 "
+
+    assert len(engine_core.scheduler.waiting) == 0
+    assert len(engine_core.scheduler.running) == 1
+
+
+def create_random_request(
+        request_id: int, num_tokens: int,
+        sampling_params: SamplingParams) -> EngineCoreRequest:
+    return EngineCoreRequest(
+        request_id=str(request_id),
+        prompt_token_ids=[request_id] * num_tokens,
+        mm_inputs=None,
+        mm_hashes=None,
+        mm_placeholders=None,
+        sampling_params=sampling_params,
+        eos_token_id=None,
+        arrival_time=0,
+        lora_request=None,
+    )
+
+
+def get_params_test_blocks_borders_aligned_prompts():
+    seqs_max_tokens = [65, 67, 7]
+    prompts_lengths = [49, 41, 7]
+    steps_add_reqs = [0, 0, 0]  # add all requests in the beginning
+
+    checked_steps = [{
+        "step": 0,
+        "tkv": 0,
+        "num_waiting": 3,
+        "num_running": 0,
+        "request_outputs": []
+    }, {
+        "step": 1,  # Prefill sequence 0
+        "tkv": 64,
+        "num_waiting": 2,
+        "num_running": 1,
+        "request_outputs": ["0"]  
+    }, {
+        "step": 2,  # Prefill sequence 1
+        "tkv": 64,  # Still 64 because this step is also a prefill
+        "num_waiting": 1,
+        "num_running": 2,
+        "request_outputs": ["1"]  
+    }, {
+        "step": 3,
+        "tkv": 65,  # Two decodes increases the tkv
+        "num_waiting": 1,
+        "num_running": 2,
+        "request_outputs": ["1", "0"]  # Two sequences are decoded
+    }, {
+        "step": 4,  # Check normal decode continuation
+        "tkv": 66,
+        "num_waiting": 1,
+        "num_running": 2,
+        "request_outputs": ["1", "0"]
+    }, {
+        "step": 65,  # Last step before fist sequence finishes
+        "tkv": 127,
+        "num_waiting": 1,
+        "num_running": 2,
+        "request_outputs": ["1", "0"]
+    # }, {
+    #     "step": 66,  # Sequence 0 finishes at step 66 (2 prefills + 64 decodes)
+    #     "tkv": 128,
+    #     "num_waiting": 1,
+    #     "num_running": 1,
+    #     "request_outputs": ["1"]  # TODO this fails?? Why do we observe ["1", "0"]?
```

Comment:
I think that's because when a request finishes, it is marked `finished` at the last step. So here, in step 66 so it still shows up in `request_outputs` as `finished` and then gets cleared in the next step. `request_outputs[1].finished` should be `True` in this case

Reply:
That should happen for request 1 as well though when it finishes 🤔

Reply:
Worth exploring if we can attach `request_outputs[n].finished` as a boolean check for each step somehow

Reply:
good idea, maybe test also which ids are running and waiting instead of the length. This will help to decorticate the steps